File: projects/ruby/lib/basic_files.rb
# frozen_string_literal: true

require "date"

Task = Struct.new(:title, :status, :due, keyword_init: true)

module BasicFiles
  module_function

  def summarize(tasks)
    tasks.each_with_object(Hash.new(0)) do |task, counts|
      counts[task.status] += 1
    end
  end

  def label(task)
    case task.status
    when :done
      "#{task.title}: complete"
    else
      "#{task.title}: #{task.status}"
    end
  end
end

if $PROGRAM_NAME == __FILE__
  tasks = [
    Task.new(title: "Add Ruby fixture", status: :doing, due: Date.new(2026, 5, 27)),
    Task.new(title: "Try syntax themes", status: :todo, due: nil)
  ]
  puts BasicFiles.summarize(tasks)
end

File: projects/python/src/basic_files.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from enum import StrEnum


class Status(StrEnum):
    TODO = "todo"
    DOING = "doing"
    DONE = "done"


@dataclass(frozen=True)
class Task:
    title: str
    status: Status
    due: date | None = None

    @property
    def is_open(self) -> bool:
        return self.status is not Status.DONE


def summarize(tasks: list[Task]) -> dict[Status, int]:
    counts = {status: 0 for status in Status}
    for task in tasks:
        counts[task.status] += 1
    return counts


if __name__ == "__main__":
    backlog = [
        Task("Collect sample files", Status.DOING, date(2026, 5, 27)),
        Task("Review syntax output", Status.TODO),
    ]
    print(summarize(backlog))
